File: brick_sim_js/js/components/item.js
/**
 * Brick Simulation Game - Item Component
 * Handles power-up items dropped by broken bricks
 */

class Item extends GameObject {
    /**
     * Creates a new item
     * @param {number} x - The x-coordinate
     * @param {number} y - The y-coordinate
     * @param {number} size - The size of the item
     * @param {string} type - The type of item
     * @param {number} speed - The falling speed of the item
     * @param {number} screenHeight - The height of the game screen
     */
    constructor(x, y, size, type, speed, screenHeight) {
        let color;
        
        // Set color based on item type
        switch (type) {
            case 'extend':
                color = '#00FF00'; // Green
                break;
            case 'slow':
                color = '#00FFFF'; // Cyan
                break;
            case 'multi':
                color = '#FF00FF'; // Magenta
                break;
            case 'life':
                color = '#FF0000'; // Red
                break;
            case 'laser':
                color = '#FFFF00'; // Yellow
                break;
            case 'fast':
                color = '#FFA500'; // Orange
                break;
            case 'warp':
                color = '#9932CC'; // Purple
                break;
            default:
                color = '#FFFFFF'; // White
        }
        
        super(x, y, size, size, color);
        
        this.type = type;
        this.velocityY = speed;
        this.screenHeight = screenHeight;
        this.collected = false;
        this.animationTime = 0;
        this.blinkRate = 200; // milliseconds
        this.rotationAngle = 0;
        this.symbolScale = 0.6; // Scale of the symbol relative to item size
    }
    
    /**
     * Updates the item state
     * @param {number} deltaTime - The time passed since the last update in milliseconds
     */
    update(deltaTime) {
        if (!this.active || this.collected) return;
        
        // Update animation time for blinking effect
        this.animationTime += deltaTime;
        
        // Update rotation angle for spinning effect
        this.rotationAngle = (this.rotationAngle + 0.05) % (Math.PI * 2);
        
        // Update position
        super.update(deltaTime);
        
        // Check if item is out of bounds (bottom of screen)
        if (this.y > this.screenHeight) {
            this.active = false;
            this.visible = false;
        }
    }
    
    /**
     * Collects the item
     * @returns {string} The type of item collected
     */
    collect() {
        if (!this.active || this.collected) return null;
        
        this.collected = true;
        this.active = false;
        this.visible = false;
        
        return this.type;
    }
    
    /**
     * Renders the item with visual effects
     * @param {CanvasRenderingContext2D} ctx - The canvas rendering context
     */
    render(ctx) {
        if (!this.visible || this.collected) return;
        
        if (this.sprite) {
            // Render sprite if available
            ctx.drawImage(this.sprite, this.x, this.y, this.width, this.height);
        } else {
            ctx.save();
            
            // Make items blink by alternating opacity
            const blinkState = Math.floor(this.animationTime / this.blinkRate) % 2;
            ctx.globalAlpha = blinkState ? 0.7 : 1;
            
            // Draw item background
            ctx.fillStyle = this.color;
            ctx.beginPath();
            ctx.arc(this.x + this.width / 2, this.y + this.height / 2, this.width / 2, 0, Math.PI * 2);
            ctx.fill();
            
            // Draw white border
            ctx.strokeStyle = '#FFFFFF';
            ctx.lineWidth = 2;
            ctx.beginPath();
            ctx.arc(this.x + this.width / 2, this.y + this.height / 2, this.width / 2 - 1, 0, Math.PI * 2);
            ctx.stroke();
            
            // Draw symbol based on item type
            ctx.translate(this.x + this.width / 2, this.y + this.height / 2);
            ctx.rotate(this.rotationAngle);
            ctx.fillStyle = '#FFFFFF';
            
            const symbolSize = this.width * this.symbolScale;
            
            switch (this.type) {
                case 'extend':
                    // Draw a horizontal bar (paddle extension)
                    ctx.fillRect(-symbolSize / 2, -symbolSize / 6, symbolSize, symbolSize / 3);
                    break;
                    
                case 'slow':
                    // Draw an arrow pointing down
                    ctx.beginPath();
                    ctx.moveTo(0, symbolSize / 2);
                    ctx.lineTo(-symbolSize / 2, -symbolSize / 2);
                    ctx.lineTo(symbolSize / 2, -symbolSize / 2);
                    ctx.closePath();
                    ctx.fill();
                    break;
                    
                case 'multi':
                    // Draw multiple small circles (multi-ball)
                    const circleRadius = symbolSize / 6;
                    for (let angle = 0; angle < Math.PI * 2; angle += Math.PI / 2) {
                        const circleX = Math.cos(angle) * (symbolSize / 2 - circleRadius);
                        const circleY = Math.sin(angle) * (symbolSize / 2 - circleRadius);
                        ctx.beginPath();
                        ctx.arc(circleX, circleY, circleRadius, 0, Math.PI * 2);
                        ctx.fill();
                    }
                    ctx.beginPath();
                    ctx.arc(0, 0, circleRadius, 0, Math.PI * 2);
                    ctx.fill();
                    break;
                    
                case 'life':
                    // Draw a heart
                    ctx.beginPath();
                    const heartSize = symbolSize / 2;
                    ctx.moveTo(0, heartSize / 4);
                    ctx.bezierCurveTo(heartSize / 2, -heartSize / 2, heartSize, 0, 0, heartSize);
                    ctx.bezierCurveTo(-heartSize, 0, -heartSize / 2, -heartSize / 2, 0, heartSize / 4);
                    ctx.fill();
                    break;
                    
                case 'laser':
                    // Draw a laser/lightning bolt
                    ctx.beginPath();
                    ctx.moveTo(-symbolSize / 2, -symbolSize / 2);
                    ctx.lineTo(0, 0);
                    ctx.lineTo(-symbolSize / 4, 0);
                    ctx.lineTo(symbolSize / 2, symbolSize / 2);
                    ctx.lineTo(0, 0);
                    ctx.lineTo(symbolSize / 4, 0);
                    ctx.closePath();
                    ctx.fill();
                    break;
                    
                case 'fast':
                    // Draw an arrow pointing up
                    ctx.beginPath();
                    ctx.moveTo(0, -symbolSize / 2);
                    ctx.lineTo(-symbolSize / 2, symbolSize / 2);
                    ctx.lineTo(symbolSize / 2, symbolSize / 2);
                    ctx.closePath();
                    ctx.fill();
                    break;
                    
                case 'warp':
                    // Draw a star
                    ctx.beginPath();
                    const spikes = 5;
                    const outerRadius = symbolSize / 2;
                    const innerRadius = symbolSize / 4;
                    
                    for (let i = 0; i < spikes * 2; i++) {
                        const radius = i % 2 === 0 ? outerRadius : innerRadius;
                        const angle = (i * Math.PI) / spikes;
                        const x = Math.cos(angle) * radius;
                        const y = Math.sin(angle) * radius;
                        
                        if (i === 0) {
                            ctx.moveTo(x, y);
                        } else {
                            ctx.lineTo(x, y);
                        }
                    }
                    ctx.closePath();
                    ctx.fill();
                    break;
            }
            
            ctx.restore();
        }
    }
}

File: brick_sim_js/src/components/item.py
"""
Item component - power-ups that can be collected by the player
"""
import pygame
from .game_object import GameObject
from ..utils.helpers import load_image

class Item(GameObject):
    """
    Item class represents power-ups that fall from destroyed bricks
    and can be caught by the paddle for various effects
    """
    def __init__(self, x, y, size, item_type, color, speed):
        super().__init__(x, y, size, size, color)
        self.item_type = item_type
        self.speed = speed
        # Try to load item sprite based on type (falls back to shape if not found)
        self.sprite = load_image(f"items/{item_type}.png")
        # Set symbol for fallback rendering
        self.symbol = self._get_symbol_for_type()
        
    def _get_symbol_for_type(self):
        """Get a symbol to display inside the item based on its type"""
        symbols = {
            "extend": "+",    # Extend paddle
            "slow": "S",      # Slow ball
            "multi": "M",     # Multiple balls
            "life": "♥",      # Extra life
            "laser": "L",     # Laser
            "fast": "F",      # Speed up
            "warp": "W",      # Warp to next level
        }
        return symbols.get(self.item_type, "?")
        
    def update(self):
        """Update item position (falling down)"""
        self.y += self.speed
        
    def render(self, screen):
        """Render the item using sprite or shape"""
        if self.sprite:
            # If we have a sprite, scale it to match item dimensions and draw it
            scaled_sprite = pygame.transform.scale(self.sprite, (self.width, self.height))
            screen.blit(scaled_sprite, (self.x, self.y))
        else:
            # Otherwise, draw a colored square with a symbol
            self.draw_shape(screen)
            
            # Add symbol text in the center
            font = pygame.font.SysFont(None, int(self.width * 0.8))
            text = font.render(self.symbol, True, (255, 255, 255))
            text_rect = text.get_rect(center=(self.x + self.width // 2, self.y + self.height // 2))
            screen.blit(text, text_rect)
    
    def apply_effect(self, game):
        """Apply the item's effect to the game"""
        if self.item_type == "extend":
            # Extend paddle
            game.paddle.extend()
            game.schedule_event(10000, game.paddle.reset_size)  # Reset after 10 sec
            
        elif self.item_type == "slow":
            # Slow down ball(s)
            for ball in game.balls:
                ball.decrease_speed(0.7)
            game.schedule_event(15000, lambda: [ball.increase_speed(1/0.7) for ball in game.balls])
            
        elif self.item_type == "multi":
            # Add 2 extra balls
            if len(game.balls) < 5:  # Limit total number of balls
                current_ball = game.balls[0]
                for angle in [30, -30]:
                    new_ball = game.create_ball(current_ball.x, current_ball.y)
                    new_ball.is_launched = True
                    new_ball.set_direction(angle)
                    game.balls.append(new_ball)
                    
        elif self.item_type == "life":
            # Add extra life
            game.lives += 1
            
        elif self.item_type == "laser":
            # Enable laser mode
            game.enable_laser_mode(10000)  # 10 seconds
            
        elif self.item_type == "fast":
            # Speed up game pace
            for ball in game.balls:
                ball.increase_speed(1.3)
            game.schedule_event(15000, lambda: [ball.decrease_speed(1/1.3) for ball in game.balls])
            
        elif self.item_type == "warp":
            # Skip to next level
            game.next_level()
        
        # Deactivate the item after applying its effect
        self.deactivate()
